cmdstanpy/model.py

```diff
@@ -133,85 +132,88 @@ def compile(
-            hpp_file = Path(hpp_file).as_posix()
-            if overwrite or not os.path.exists(hpp_file):
-                self._logger.info('stan to c++ (%s)', hpp_file)
-                stanc_path = os.path.join(
-                    cmdstan_path(), 'bin', 'stanc' + EXTENSION
-                )
-                stanc_path = Path(stanc_path).as_posix()
-                cmd = [
-                    stanc_path,
-                    '--o={}'.format(hpp_file),
-                    Path(stan_file).as_posix(),
-                ]
-                if include_paths is not None:
-                    bad_paths = [
-                        d for d in include_paths if not os.path.exists(d)
-                    ]
-                    if any(bad_paths):
-                        raise Exception(
-                            'invalid include paths: {}'.format(
-                                ', '.join(bad_paths)
-                            )
-                        )
-                    cmd.append(
-                        '--include_paths='
-                        + ','.join((Path(p).as_posix() for p in include_paths))
-                    )
-
-                do_command(cmd, logger=self._logger)
-                if not os.path.exists(hpp_file):
-                    raise Exception('syntax error'.format(stan_file))
-
             exe_file, _ = os.path.splitext(os.path.abspath(stan_file))
             exe_file = Path(exe_file).as_posix()
             exe_file += EXTENSION
-            make = os.getenv('MAKE', 'make')
-            cmd = [make, 'O={}'.format(opt_lvl), exe_file]
-            self._logger.info('compiling c++')
-            try:
-                do_command(cmd, cmdstan_path(), self._logger)
-            except Exception as e:
-                self._logger.error('make cmd failed %s', e)
-                compilation_failed = True
-            if is_copied and not compilation_failed:
-
-                original_target_dir = os.path.dirname(self._stan_file)
-                # reconstruct the output file name
-                new_exec_name = (
-                    os.path.basename(os.path.splitext(self._stan_file)[0])
-                    + EXTENSION
-                )
-
-                self._exe_file = os.path.join(
-                    original_target_dir, new_exec_name
-                )
-                # copy the generated file back to the original directory
-                shutil.copy(exe_file, self._exe_file)
-            elif not compilation_failed:
-                self._exe_file = exe_file
-        if compilation_failed:
-            self._logger.error('model compilation failed')
-        else:
-            self._logger.info('compiled model file: %s', self._exe_file)
+            do_compile = True
+            if os.path.exists(exe_file):
+                src_time = os.path.getmtime(self._stan_file)
+                exe_time = os.path.getmtime(exe_file)
+                if exe_time > src_time:
```

Comment:
I think this will work most times, unless user use the copy without metadata.

But then, the worstcase is that model is compiled once.